### PR TITLE
feat(core): add pi event adapter for 0.41 ingest

### PR DESCRIPTION
**Pull request by ZeR020**
Second PR in the pi integration stack — context and overall design in #1430. Builds on #1473 (merged); the diff here is the pi adapter only.

## Why

Pi events need a core adapter that produces the same envelope shape `ingestRawEvents` already accepts, with `source: "pi"` explicit and deterministic ids so HTTP/CLI retries do not duplicate rows.

## What

- `packages/core/src/pi-hooks.ts`: map session/prompt/tool events; compaction is flush-only
- Envelope fields include `session_stream_id` / `session_id` aliases required by 0.41 ingest
- Test: `ingestRawEvents` persists source `pi` and creates zero opencode rows

## Test plan

- `vitest run packages/core/src/pi-hooks.test.ts` — 40/40 passed
- Gates `lint`, `tsc`, `build` green locally; `test` passes except local-sandbox env failures unrelated to this change (sqlite-vec native ext unavailable in this dev sandbox)

@kunickiaj layer 2 of the stack is ready for review whenever you have a moment. Layers 3–8 will follow one at a time as each predecessor merges, same as this one.

Related: #1430, #1429, #1473

### CODE REVIEW

File: packages/core/src/pi-hooks.ts [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+		const reason = field(payload, "reason");
+		eventType = "session_end";
+		eventPayload = { reason: reason ?? null };
+		idPart = entryId || "session_end";
+		consumed.add("reason");
+	} else if (piEvent === "message_end") {
+		const role = coerceString(field(payload, "role")).toLowerCase();
+		const text = coerceString(field(payload, "text", "content", "prompt"));
+		if (!text) return null;
+		if (!entryId) return null;
```
[the remark this answers is not in the document]
Comment (ZeR020) [reply packages/core/src/pi-hooks.ts:190]: Correct about Pi 0.41's native `message_end` shape — but this adapter intentionally never sees the raw pi event. The pi extension's `pi.on("message_end")` handler unwraps `event.message` (role/text via `extractMessageRole`/`extractMessageText`) and synthesizes a deterministic `entryId` (`stableMessageEntryId`, timestamp-discriminated) before POSTing the flat payload this mapper consumes (`packages/pi-extension/src/index.ts`, `payloads.ts` — landing in a later layer of this stack, see #1430). Keeping the unwrap on the producer side is what lets core stay pi-agnostic. Traced end-to-end: no caller passes the raw nested event. No change needed here.

File: packages/core/src/pi-hooks.ts [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+	return false;
+}
+
+/**
+ * Deterministic event id: `pi:<sessionId>:<entryId|toolCallId|stable-suffix>`.
+ * Same logical event must yield the same id across HTTP/CLI retries.
+ */
+function buildPiEventId(sessionId: string, stablePart: string): string {
+	const part = stablePart.trim();
+	if (!part) throw new Error("pi event id stable part is required");
```
Comment (kunickiaj) [inline packages/core/src/pi-hooks.ts]: Raw ingestion validates event_id against /^[A-Za-z0-9._:-]+$/ and a 128-character limit, but this concatenates unconstrained session and tool/message IDs. For example, sessionId="session/with/slash" and entryId="entry+1" produces an ID that ingestRawEvents rejects with RawEventIngestValidationError. Pi toolCallId is typed only as string, so provider-specific IDs can trigger this even if session IDs are UUIDs. Please hash the identity components into a fixed-length safe ID, as the Claude and Codex adapters do, and add charset/length regression coverage.

File: packages/core/src/pi-hooks.ts [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+
+	if (piEvent === "session_start") {
+		eventType = "session_start";
+		eventPayload = {};
+		idPart = entryId || "session_start";
+	} else if (piEvent === "session_shutdown") {
+		const reason = field(payload, "reason");
+		eventType = "session_end";
+		eventPayload = { reason: reason ?? null };
+		idPart = entryId || "session_end";
```
Comment (kunickiaj) [inline packages/core/src/pi-hooks.ts]: This fallback gives every shutdown for a session the same event ID. Pi can shut an extension down for reload and later shut the same session down for exit; the second event is then deduplicated, preserving the earlier reason/timestamp and losing the final lifecycle event. The continuation stack's payload builder also currently supplies the fixed entryId "session_end", so it hits the same collision. Please require or derive a retry-stable discriminator for each distinct shutdown occurrence and add a reload-then-exit regression test.

**Review by kunickiaj — CHANGES_REQUESTED** [review]
Two ingestion-identity issues need correction before this adapter is safe to merge.

File: packages/core/src/pi-hooks.ts (hunk 2 above)
Comment (ZeR020) [reply packages/core/src/pi-hooks.ts]: Hashed identity into `pi_evt_` + sha256(sessionId|piEvent|idPart).hex.slice(0, 24), matching Claude/Codex. Added charset/length coverage, including `session/with/slash` + `entry+1`.

05d7b2f7

File: packages/core/src/pi-hooks.ts (hunk 3 above)
Comment (ZeR020) [reply packages/core/src/pi-hooks.ts]: Shutdown no longer uses a fixed `session_end` id. When `entryId` is missing or that generic value, the discriminator is `session_end:<reason>`. Reload then quit persist as two events; identical retries still dedupe.

05d7b2f7

**Comment by ZeR020** [comment]
@kunickiaj requested identity fixes are in `05d7b2f7` (hashed event IDs + shutdown discriminator). CI is green. Ready for re-review.

**Comment by ZeR020** [comment]
@kunickiaj ready for your review whenever you have a moment.

File: packages/core/src/pi-hooks.ts [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+
+	if (piEvent === "session_start") {
+		eventType = "session_start";
+		eventPayload = {};
+		idPart = entryId || "session_start";
+	} else if (piEvent === "session_shutdown") {
+		const reason = field(payload, "reason");
+		eventType = "session_end";
+		eventPayload = { reason: reason ?? null };
+		idPart = entryId && entryId !== "session_end" ? entryId : `session_end:${coerceString(reason)}`;
```
Comment (kunickiaj) [inline packages/core/src/pi-hooks.ts]: Two shutdowns with the same reason (for example, `/reload` twice) still produce the same ID, so the second is dropped. Please have the producer send a stable per-occurrence `entryId`—or use its stable payload timestamp as the fallback—and add reload → reload coverage.

File: packages/core/src/pi-hooks.ts [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+		consumed.add("is_error");
+		consumed.add("error");
+		consumed.add("tool_error");
+	} else {
+		return null;
+	}
+
+	if (!idPart) return null;
+
+	const meta: Record<string, unknown> = {
```
Comment (kunickiaj) [inline packages/core/src/pi-hooks.ts:287]: Please match the Claude and Codex adapters by exporting `PI_EVENT_ID_ALGO = "pi/1"` and storing it as `meta.event_id_algo`. This keeps stored IDs attributable if the algorithm changes.

File: packages/core/src/pi-hooks.test.ts [hunk 6]
```diff
@@ -0,0 +1,824 @@
...
+			}),
+		);
+		expect(envelope.event_id).toMatch(/^[A-Za-z0-9._:-]+$/);
+		expect(envelope.event_id.length).toBeLessThanOrEqual(128);
+		const result = ingestRawEvents(store, envelope);
+		expect(result.inserted).toBe(1);
+		expect(result.skipped).toBe(0);
+	});
+
+	it("persists reload then exit as distinct session_end events", () => {
```
Comment (kunickiaj) [inline packages/core/src/pi-hooks.test.ts:804]: This covers reload → quit, whose reasons already differ. Please add reload → reload: both occurrences should insert, while retrying either payload should skip.

**Review by kunickiaj — CHANGES_REQUESTED** [review]
Thanks for addressing the original review. The hashed-ID fix looks good. One shutdown edge case still blocks approval; the other two comments are small identity-contract follow-ups that should fit naturally with that change.

File: packages/core/src/pi-hooks.test.ts (hunk 6 above)
Comment (ZeR020) [reply packages/core/src/pi-hooks.test.ts:804]: Added reload → reload coverage: both insert; retry of either payload skips.

File: packages/core/src/pi-hooks.ts (hunk 4 above)
Comment (ZeR020) [reply packages/core/src/pi-hooks.ts]: Generic `session_end` identity now includes the payload timestamp. Two `/reload` events with different `ts` insert as separate rows; retrying the same payload still skips. Unique `entryId` still wins when present.

Anything else needed on this review?

File: packages/core/src/pi-hooks.ts (hunk 5 above)
Comment (ZeR020) [reply packages/core/src/pi-hooks.ts:287]: Exported `PI_EVENT_ID_ALGO = "pi/1"` and stored it as `meta.event_id_algo`.

**Comment by ZeR020** [comment]
@kunickiaj ready for your review whenever you have a moment. Second-round items are in (per-occurrence shutdown IDs, `pi/1` algo tag, reload→reload coverage).

GitHub Actions is green; Snyk is failing on this org's private-test quota.

**Comment by ZeR020** [comment]
@kunickiaj whenever you have a moment, could you please take another look at this PR? I’ve addressed the review feedback from the previous round and pushed the fixes. Would appreciate a re-review when you get a chance.

**Review by kunickiaj — APPROVED** [review]
